### PR TITLE
feat: Use specific images for SMP v2 jobs

### PR DESCRIPTION
**Pull request by Teng-xu**
*Issue #, if available:*

*Description of changes:*
From v2, SMP will have their own Pytorch docker images. This PR ensures they are used.
*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [ ] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [ ] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/image_uris.py [hunk 1]
```diff
@@ -660,6 +660,12 @@ def get_training_image_uri(
         container_version = None
         base_framework_version = None
 
+    # Check for smp library
+    if distribution is not None:
+        if "torch_distributed" in distribution and "smdistributed" in distribution:
+            if "modelparallel" in distribution["smdistributed"]:
+                framework = "pytorch-smp"
```
Comment (rahul003) [inline src/sagemaker/image_uris.py:667]: should we check enabled or True values?

Comment (Teng-xu) [reply src/sagemaker/image_uris.py:667]: I think enabled = False is not really a common use case, if the user specifies these fields then they should be intended to use our lib, so we don't need to check that, what do you think?

Comment (rahul003) [reply src/sagemaker/image_uris.py:667]: do other checks in sdk do that?

Comment (rahul003) [reply src/sagemaker/image_uris.py:667]: i think we should check

Comment (Teng-xu) [reply src/sagemaker/image_uris.py:667]: okay I will add that

**Review by akrishna1995 — COMMENTED** [review]
/bot run all

File: src/sagemaker/image_uris.py (hunk 1 above)
Comment (akrishna1995) [reply src/sagemaker/image_uris.py]: @Teng-xu @rahul003 - is this comment resolved ?

Comment (rahul003) [reply src/sagemaker/image_uris.py]: Yes

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-unit-tests
* Commit ID: 3e73f7120d93b3ef8b473b153ce045a254e8f7e7
* Result: FAILED
* [Build Logs](https://fegcnxu6da.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=38c845d1-ead0-4b82-a586-5e33b680c185%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by akrishna1995** [comment]
@Teng-xu can you run  black-check, flake8, pylint 

https://github.com/aws/sagemaker-python-sdk/blob/master/.githooks/pre-push 

```
=================================== log end ====================================
âœ– FAIL pylint in 4 minutes, 34.53 seconds
___________________________________ summary ____________________________________
ERROR:   flake8: parallel child exit code 1
ERROR:   pylint: parallel child exit code 1
  docstyle: commands succeeded
ERROR:   black-check: parallel child exit code 1
  twine: commands succeeded

```

**Comment by akrishna1995** [comment]
> @Teng-xu can you run black-check, flake8, pylint
> 
> https://github.com/aws/sagemaker-python-sdk/blob/master/.githooks/pre-push
> 
> ```
> =================================== log end ====================================
> âœ– FAIL pylint in 4 minutes, 34.53 seconds
> ___________________________________ summary ____________________________________
> ERROR:   flake8: parallel child exit code 1
> ERROR:   pylint: parallel child exit code 1
>   docstyle: commands succeeded
> ERROR:   black-check: parallel child exit code 1
>   twine: commands succeeded
> ```

@Teng-xu @rahul003

File: src/sagemaker/image_uris.py [hunk 2]
```diff
@@ -660,6 +660,17 @@ def get_training_image_uri(
         container_version = None
         base_framework_version = None
 
+    # Check for smp library
+    if distribution is not None:
+        if "torch_distributed" in distribution and "smdistributed" in distribution:
+            if "modelparallel" in distribution["smdistributed"]:
+                if distribution["smdistributed"]["modelparallel"].get("enabled", True):
+                    framework = "pytorch-smp"
+                    if "p5" in instance_type:
```
Comment (sheilaliuxl) [inline src/sagemaker/image_uris.py:669]: We need tests to cover 1) both cases and 2) SMP not enabled case.

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-local-mode-tests
* Commit ID: 3e73f7120d93b3ef8b473b153ce045a254e8f7e7
* Result: SUCCEEDED
* [Build Logs](https://8huhkqo601.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=93d0f28e-4b74-4cc2-904d-fa4e5a09f166%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-notebook-tests
* Commit ID: 3e73f7120d93b3ef8b473b153ce045a254e8f7e7
* Result: SUCCEEDED
* [Build Logs](https://bkqkixfb59.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=9022fa76-b5a2-4003-9bc3-5f06b6e892d3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-slow-tests
* Commit ID: 3e73f7120d93b3ef8b473b153ce045a254e8f7e7
* Result: SUCCEEDED
* [Build Logs](https://omfrydptb9.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=164f8a6d-01c4-45aa-a248-691af1bc44a2%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-pr
* Commit ID: 3e73f7120d93b3ef8b473b153ce045a254e8f7e7
* Result: SUCCEEDED
* [Build Logs](https://abkt2mlh3k.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=13f42bd8-2749-4c7d-8454-9fde95723eb0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Review by akrishna1995 — COMMENTED** [review]
/bot run all

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-local-mode-tests
* Commit ID: 1344b23eaef6823d4ca8b9ed5867164256eb4068
* Result: SUCCEEDED
* [Build Logs](https://8huhkqo601.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=46e20638-743b-4da3-a5bf-04a1fcfe4cab%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-notebook-tests
* Commit ID: 1344b23eaef6823d4ca8b9ed5867164256eb4068
* Result: SUCCEEDED
* [Build Logs](https://bkqkixfb59.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=c5b868f3-749e-4888-a72a-62d48a835392%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-pr
* Commit ID: 1344b23eaef6823d4ca8b9ed5867164256eb4068
* Result: SUCCEEDED
* [Build Logs](https://abkt2mlh3k.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=7ab3371e-c465-4789-9433-ced3738aa7af%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-slow-tests
* Commit ID: 1344b23eaef6823d4ca8b9ed5867164256eb4068
* Result: SUCCEEDED
* [Build Logs](https://omfrydptb9.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=3b60574f-0365-4bf3-926e-1c09ee6d44af%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Review by akrishna1995 — APPROVED** [review]
approved - UTs have passed locally

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-unit-tests
* Commit ID: 1344b23eaef6823d4ca8b9ed5867164256eb4068
* Result: SUCCEEDED
* [Build Logs](https://fegcnxu6da.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=d2853011-db91-48f7-b0a4-591c798dc98e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: tests/unit/sagemaker/image_uris/test_smp_v2.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+from __future__ import absolute_import
+
+import pytest
+from sagemaker import image_uris
+from tests.unit.sagemaker.image_uris import expected_uris
+
+CONTAINER_VERSIONS = {"ml.p4d.24xlarge": "cu118", "ml.p5d.24xlarge": "cu121"}
```
Comment (sheilaliuxl) [inline tests/unit/sagemaker/image_uris/test_smp_v2.py:19]: Here it has a typo, I believe it should be `"ml.p5d.24xlarge"`:
- See https://aws.amazon.com/ec2/instance-types/

**Pull request merged**